snapdtool/fips_linux.go

```diff
@@ -0,0 +1,146 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapdtool
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/fips"
+	"github.com/snapcore/snapd/release"
+)
+
+func findFIPSLibsAndModules(snapRoot string) (opensslLib, module string) {
+	// with OpenSSL supported by 22.04 (the snapd snap build base), the FIPS
+	// module is available as an *.so library loaded by libcrypto.so.3 at
+	// runtime.
+	var fipsLibAndModulePathInSnapdSnap = []struct {
+		opensslLib, moduleFile string
+	}{
+		{"usr/lib/x86_64-linux-gnu/libcrypto.so.3", "usr/lib/x86_64-linux-gnu/ossl-modules-3/fips.so"},
+		{"usr/lib/aarch64-linux-gnu/libcrypto.so.3", "usr/lib/aarch64-linux-gnu/ossl-modules-3/fips.so"},
+		{"usr/lib/arm-linux-gnueabihf/libcrypto.so.3", "usr/lib/arm-linux-gnueabihf/ossl-modules-3/fips.so"},
+		{"usr/lib/i386-linux-gnu/libcrypto.so.3", "usr/lib/i386-linux-gnu/ossl-modules-3/fips.so"},
+		{"usr/lib/riscv64-linux-gnu/libcrypto.so.3", "usr/lib/riscv64-linux-gnu/ossl-modules-3/fips.so"},
+		{"usr/lib/s390x-linux-gnu/libcrypto.so.3", "usr/lib/s390x-linux-gnu/ossl-modules-3/fips.so"},
+	}
+
+	for _, bundle := range fipsLibAndModulePathInSnapdSnap {
+		lib := filepath.Join(snapRoot, bundle.opensslLib)
+		module := filepath.Join(snapRoot, bundle.moduleFile)
+		if osutil.FileExists(lib) && osutil.FileExists(module) {
+			return lib, module
+		}
+	}
+	return "", ""
+}
+
+// MaybeSetupFIPS checks whether a system-wide FIPS mode is enabled and if
+// so sets up an environment such that the current process is able to use
+// libraries that are required for FIPS compliance and reexecs.
+func MaybeSetupFIPS() error {
+	enabled, err := fips.IsEnabled()
+	if err != nil {
+		return fmt.Errorf("cannot obtain FIPS status: %w", err)
+	}
+
+	if !enabled {
+		// FIPS not enabled do nothing
+		return nil
+	}
+
+	logger.Debugf("FIPS mode enabled system wide")
+
+	if os.Getenv("SNAPD_FIPS_BOOTSTRAP_DONE") == "1" {
+		// we've already been reexeced into FIPS mode and bootstrap was
+		// performed
+		logger.Debugf("FIPS bootstrap complete")
+
+		// if we reached this place, then the initialization was
+		// completed successfully and we can drop the environment
+		// variables, other processes which may be invoked by snapd will
+		// perform the initialization cycle on their own when needed
+		os.Unsetenv("GOFIPS")
+		os.Unsetenv("SNAPD_FIPS_BOOTSTRAP_DONE")
+		os.Unsetenv("OPENSSL_MODULES")
+		os.Unsetenv("GO_OPENSSL_VERSION_OVERRIDE")
```

Comment:
I feel like we should verify that initialization was done. If for some reason, some library changes a NEEDED to a dlopen, then some initialization will be delayed. It will be a well hidden bug then.

Could we somehow query go's openssl binding for where the modules are load from for example at this point?

Reply:
the way the toolchain bits work, with GOFIPS=1 there would be a panic during startup already, so we'd never reach this place, see https://github.com/microsoft/go/blob/77f45f8554027fe20fb411b40328192354e58703/patches/0004-Add-OpenSSL-crypto-backend.patch#L279